services/ansibler/server/loadbalancers.go

```diff
@@ -40,106 +40,316 @@ const (
+			for _, role := range lb.CurrentLbCluster.Roles {
+				if role.RoleType == pb.RoleType_ApiServer {
+					oldAPIServer = lb
+					break
 				}
-				return nil
-			})
-		}(lbInfo, k8sClusterName)
+			}
+			if oldAPIServer != nil {
+				break
+			}
+		}
+
+		if oldAPIServer != nil {
+			oldEndpoint = oldAPIServer.CurrentLbCluster.Dns.Endpoint
+			break
+		}
+
+		// 2nd pick the control node as the previous ApiServer.
+		node, err := FindAPIEndpointNode(k8sCluster.TargetK8sNodepool)
+		if err != nil {
+			return fmt.Errorf("failed to find ApiEndpoint k8s node, couldn't update Api server endpoint")
+		}
+
+		oldEndpoint = node.Public
+	case AttachingLoadBalancer:
+		node, err := FindAPIEndpointNode(k8sCluster.TargetK8sNodepool)
+		if err != nil {
+			// If no Node has type ApiEndpoint this means that the cluster
+			// wasn't build yet (i.e. it's the first time the manifest goes
+			// through the workflow), thus we don't need to change the api endpoint.
+			return nil
+		}
+
+		oldEndpoint = node.Public
+		newEndpoint = apiServer.DesiredLbCluster.Dns.Endpoint
+	case DetachingLoadBalancer:
+		// Choose one of the control nodes as the API endpoint.
+		node, err := FindAPIEndpointNode(k8sCluster.TargetK8sNodepool)
+		if err != nil {
+			return err
+		}
+
+		oldEndpoint = apiServer.CurrentLbCluster.Dns.Endpoint
+		newEndpoint = node.Public
 	}
-	err := errGroupK8s.Wait()
-	if err != nil {
-		return fmt.Errorf("error while setting up the loadbalancers : %w", err)
+
+	lbCluster := apiServer.DesiredLbCluster
+	if lbCluster == nil {
+		lbCluster = apiServer.CurrentLbCluster
+	}
+
+	log.Info().Msgf("Changing the API endpoint for the cluster %s", lbCluster.ClusterInfo.Name)
+
+	if err := changeAPIEndpoint(lbCluster.ClusterInfo.Name, oldEndpoint, newEndpoint, k8sDirectory); err != nil {
+		return fmt.Errorf("error while changing the endpoint for %s : %w", lbCluster.ClusterInfo.Name, err)
 	}
+
 	return nil
 }
 
-//changedAPIEp will change kubeadm configuration to include new EP
-//return error if not successful, nil otherwise
-func changedAPIEp(lbInfo *LBData, directory string) error {
-	log.Info().Msgf("Changing the API endpoint for the cluster %s", lbInfo.LbCluster.ClusterInfo.Name)
-	log.Info().Msgf("New endpoint is %s", lbInfo.LbCluster.Dns.Endpoint)
-	flag := fmt.Sprintf("--extra-vars \"NewEndpoint=%s OldEndpoint=%s\"", lbInfo.LbCluster.Dns.Endpoint, lbInfo.CurrentDNS.Endpoint)
-	ansible := ansible.Ansible{Playbook: apiChangePlaybook, Inventory: inventoryFile, Flags: flag, Directory: directory}
-	err := ansible.RunAnsiblePlaybook(fmt.Sprintf("EP - %s", lbInfo.LbCluster.ClusterInfo.Name))
-	if err != nil {
-		return fmt.Errorf("error while running ansible for EP change for %s : %w ", lbInfo.LbCluster.ClusterInfo.Name, err)
+// FindAPIEndpointNode searches the NodePools for a Node with type ApiEndpoint.
+// If no such node is found an error is returned.
+func FindAPIEndpointNode(nodepools []*pb.NodePool) (*pb.Node, error) {
```

Comment:
I can never decide what is better in these types of functions; return `nil` or return error :smile:.

What do you think @jaskeerat789?